pkg/logger/logger.go

```diff
@@ -96,181 +80,167 @@ func (o *LoggerOptions) apply(opts ...LoggerOption) {
-			if color {
-				enc.AppendString(TextStyle(ec.TrimmedPath(), chalk.Dim))
-			} else {
-				enc.AppendString(ec.TrimmedPath())
-			}
-		},
-		EncodeName: func(s string, enc zapcore.PrimitiveArrayEncoder) {
-			if len(s) == 0 {
-				return
-			}
-			var name string
-			if s[0] == 'x' {
-				// if the string starts with 'x', check if the name is prefixed with the
-				// dropped sample count of the form "x## " and if so, color it differently
-			LOOP:
-				for i := 1; i < len(s); i++ {
-					switch s[i] {
-					case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
-						continue
-					case ' ':
-						enc.AppendString(Color(s[:i], chalk.White))
-						name = s[i+1:]
-						break LOOP
-					default:
-						name = s
-						break LOOP
-					}
-				}
-			} else {
-				name = s
-			}
 
-			if strings.HasPrefix(name, "plugin.") {
-				enc.AppendString(Color(name, chalk.Cyan))
-			} else {
-				enc.AppendString(Color(name, chalk.Green))
-			}
-		},
-		EncodeDuration:   zapcore.SecondsDurationEncoder,
-		EncodeTime:       options.timeEncoder,
-		ConsoleSeparator: " ",
+	var middlewares []slogmulti.Middleware
+	if options.TotemFormatEnabled {
+		options.AppendName = false
+		options.Writer = os.Stderr
+		middlewares = append(middlewares, newTotemNameMiddleware())
+	}
+	if options.Sampling != nil {
+		middlewares = append(middlewares, options.Sampling.NewMiddleware())
 	}
-	level := zap.NewAtomicLevelAt(options.logLevel)
-	if options.writer != nil {
-		ws := zapcore.Lock(zapcore.AddSync(options.writer))
-		encoder := zapcore.NewConsoleEncoder(encoderConfig)
-		core := zapcore.NewCore(encoder, ws, level)
-		return &extendedSugaredLogger{
-			SugaredLogger: zap.New(core, options.zapOptions...).Sugar(),
-			level:         level,
+	var chain *slogmulti.PipeBuilder
+	for i, middleware := range middlewares {
+		if i == 0 {
+			chain = slogmulti.Pipe(middleware)
+		} else {
+			chain = chain.Pipe(middleware)
 		}
 	}
-	zapConfig := zap.Config{
-		Level:             level,
-		Development:       false,
-		DisableCaller:     options.disableCaller,
-		DisableStacktrace: true,
-		Sampling:          options.sampling,
-		Encoding:          "console",
-		EncoderConfig:     encoderConfig,
-		OutputPaths:       []string{"stderr"},
-		ErrorOutputPaths:  []string{"stderr"},
+
+	handler := newColorHandler(options.Writer, options)
+
+	if chain != nil {
+		handler = chain.Handler(handler)
 	}
-	lg, err := zapConfig.Build(options.zapOptions...)
-	if err != nil {
-		panic(err)
+
+	// write logs to a file
+	if options.RemoteSource != "" {
+		filename := logFileName + options.RemoteSource
+		f, err := logFs.OpenFile(filename, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0666)
+		if err != nil {
+			panic(err)
+		}
+		// FIXME where to close this file?
+
+		logFileHandler := newProtoHandler(f, ConfigureProtoOptions(options))
+
+		// distribute logs to handlers in parallel
+		return slogmulti.Fanout(handler, logFileHandler)
```

Comment:
`logFs.OpenFile` sometimes causes a data race when the `proto_handler` writes to the log file with `_, err = h.w.Write(*buf)` and the `seek` call in `OpenFile` [here](https://github.com/spf13/afero/blob/master/memmap.go#L262). If I remove the `os.O_APPEND` flag, not all the logs get written to the file. Any suggestions to handle the file correctly?

stack trace: https://gist.github.com/jan-law/eb20c79a98ba78f1091cf941e748e8ef

Reply:
Appends are only atomic under certain conditions. (see the section `PIPE_BUF` in `man 7 pipe` for details). You might need to change the O_NONBLOCK flag and/or write in chunks of 4096 bytes

Reply:
From the man page, it looks like writes of > 4096 bytes are not guaranteed to be atomic regardless if O_NONBLOCK is set. I ended up writing in chunks of 4096 bytes.